app/models/ci/commit.rb

```diff
@@ -125,10 +125,27 @@ def latest_statuses
       @latest_statuses ||= statuses.latest.to_a
     end
 
+    def latest_generic_statuses
+      @latest_generic_statuses ||= latest_statuses.select { |s| s.is_a?(GenericCommitStatus) }
+    end
+
+    def latest_ci_statuses
+      @latest_ci_statuses ||= latest_statuses.select { |s| s.is_a?(Ci::Build) }
+    end
+
     def latest_builds
       @latest_builds ||= builds.latest.to_a
     end
 
+    def statuses_with_matching_status
+      matching = if running? || pending?
+                   ['running', 'pending']
+                 else
+                   [status]
+                 end
+      (matching.include?(ci_status) ? 1 : 0) + latest_generic_statuses.count { |s| matching.include?(s.status) }
```

Comment:
See above for a general explanation, `running` and `pending` need to be handled that way because `Ci::Status.get_status` returns `running` if any of the statuses are either running or pending